client/src/main/java/io/pravega/client/segment/impl/ConditionalOutputStreamImpl.java

```diff
@@ -0,0 +1,139 @@
+ */
+package io.pravega.client.segment.impl;
+
+import io.netty.buffer.Unpooled;
+import io.pravega.client.netty.impl.ConnectionFactory;
+import io.pravega.client.netty.impl.RawClient;
+import io.pravega.client.stream.impl.Controller;
+import io.pravega.common.util.Retry;
+import io.pravega.common.util.Retry.RetryWithBackoff;
+import io.pravega.shared.protocol.netty.ConnectionFailedException;
+import io.pravega.shared.protocol.netty.Reply;
+import io.pravega.shared.protocol.netty.WireCommands;
+import io.pravega.shared.protocol.netty.WireCommands.AppendSetup;
+import io.pravega.shared.protocol.netty.WireCommands.ConditionalCheckFailed;
+import io.pravega.shared.protocol.netty.WireCommands.DataAppended;
+import io.pravega.shared.protocol.netty.WireCommands.SegmentIsSealed;
+import io.pravega.shared.protocol.netty.WireCommands.WrongHost;
+import java.nio.ByteBuffer;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.Lombok;
+import lombok.RequiredArgsConstructor;
+import lombok.val;
+import lombok.extern.slf4j.Slf4j;
+
+@RequiredArgsConstructor
+@Slf4j
+class ConditionalOutputStreamImpl implements ConditionalOutputStream {
+    private static final RetryWithBackoff RETRY_SCHEDULE = Retry.withExpBackoff(1, 10, 9, 30000);
+
+    private final UUID writerId;
+    private final Segment segmentId;
+    private final Controller controller;
+    private final ConnectionFactory connectionFactory;
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+    private final Object lock = new Object();
+    @GuardedBy("lock")
+    private RawClient client = null;
+    
+    private final String delegationToken;
+    private final Supplier<Long> requestIdGenerator = new AtomicLong()::incrementAndGet;
+    
+
+    @Override
+    public String getSegmentName() {
+        return segmentId.getScopedName();
+    }
+
+    @Override
+    public boolean write(ByteBuffer data, long expectedOffset) throws SegmentSealedException {
+        synchronized (lock) { //Used to preserver order.
+            long appendSequence = requestIdGenerator.get();
+            return RETRY_SCHEDULE.retryingOn(ConnectionFailedException.class)
+                    .throwingOn(SegmentSealedException.class)
+                    .run(() -> {
+                        if (client == null || client.isClosed()) {
+                            client = new RawClient(controller, connectionFactory, segmentId);
+                            long requestId = requestIdGenerator.get();
+                            log.debug("Setting up append on segment: {}", segmentId);
+                            val reply = client.sendRequest(requestId, new WireCommands.SetupAppend(requestId, writerId, segmentId.getScopedName(), delegationToken));
+                            AppendSetup appendSetup = transformAppendSetup(reply.join());
```

Comment:
I prefer not to use `join` since you could be waiting forever if something got stuck on the server or in between. Can we use `get` with a timeout, and in case of a timeout, we close the stream?

Reply:
If I call get with a timeout and the timeout fires the caller will not know if the event was written or not and has no way to find out. This is very difficult to code against. As it is written it will retry until retry exhausted, but if it sends a request and the connection remains alive it will block for a response, however long that takes. I think this is acceptable, because if we are dropping requests or replies on the floor on the server (without sending an error), that is a severe bug and we should deal it immediately rather than forcing the client to code around the race condition it creates.